pkg/storage/http/lister.go

```diff
@@ -0,0 +1,113 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"golang.org/x/net/html"
+)
+
+// List lists all versions of a module
+func (s *ModuleStore) List(ctx context.Context, module string) ([]string, error) {
+	const op errors.Op = "http.List"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+
+	baseURL := s.moduleRoot(module)
+
+	req, _ := http.NewRequest(http.MethodGet, baseURL, nil)
+	req.SetBasicAuth(s.username, s.password)
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		kind := errors.KindUnexpected
+		if resp.StatusCode == http.StatusNotFound {
+			// This is a unique case because this is actually okay! It means that there
+			// are no versions here.
+			return []string{}, nil
+		}
+		return nil, errors.E(op, kind, errors.M(module))
+	}
+
+	// Find all the links to ".mod" files within the current module
+	//
+	// Ok, so admittedly this scheme for listing versions is a little harebrained.
+	// But that said it _is_ pretty standard for directory indexes to be formatted
+	// very closely to what we expect here. I'm not settled on this but it does
+	// work surprisingly well.
```

Comment:
Doesn't seem harebrained to me! If this is what you want to go with, let's make sure to add it to our docs (on the hugo site) and explain some options for what you can do if your storage system doesn't spit this format out

Reply:
I hadn't initially approached this from the "storage plugin" angle, but I see now that it may be inconvenient/impossible to render a directory listing for certain backends.

Perhaps `List` could support two strategies for obtaining a listing, and attempt them one at a time in order:

1. Fetch a listing in the documented format from `GET /<module>/@v/list`. Athens itself won't update this file or otherwise manage it, but this endpoint may be simpler/cleaner to implement for some storage plugins than a directory listing.
2. Scrape a version list from the directory listing at `GET /<module>/@v/` as currently implemented. This option caters to storage servers based on filesystem semantics like Artifactory or `http.FileServer`.

This approach would make implementing custom storage plugins a bit nicer at the cost of slowing down `List` for cases like mine, where the server only supports the second strategy.